daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1141,54 +1136,50 @@ private[lf] object SBuiltin {
     final protected def importCid(cid: V.ContractId): SEImportValue =
       SEImportValue(typ, V.ValueContractId(cid))
     // Callback from the engine returned NotFound
-    def handleInputKeyNotFound(machine: Machine): Boolean
-    // CallBack from the engine returned a new Cid
-    def handleInputKeyFound(machine: Machine, cid: V.ContractId): Unit
+    def handleInputKeyFound(cid: V.ContractId): SExpr
     // We already saw this key, but it was undefined or was archived
-    def handleInactiveKey(machine: Machine, gkey: GlobalKey): Unit
+    def handleKeyNotFound(gkey: GlobalKey): SExpr
     // We already saw this key and it is still active
-    def handleActiveKey(machine: Machine, cid: V.ContractId): Unit
+    def handleActiveKey(cid: V.ContractId): SExpr
 
     final def handleKnownInputKey(
-        machine: Machine,
         gkey: GlobalKey,
         keyMapping: PartialTransaction.KeyMapping,
-    ): Unit =
+    ): SExpr =
       keyMapping match {
-        case PartialTransaction.KeyActive(cid) => handleActiveKey(machine, cid)
-        case PartialTransaction.KeyInactive => handleInactiveKey(machine, gkey)
+        case PartialTransaction.KeyActive(cid) => handleActiveKey(cid)
+        case PartialTransaction.KeyInactive => handleKeyNotFound(gkey)
       }
   }
 
   private[this] object KeyOperation {
     final class Fetch(override val templateId: TypeConName) extends KeyOperation {
-      override def handleInputKeyNotFound(machine: Machine): Boolean = false
-      override def handleInputKeyFound(machine: Machine, cid: V.ContractId): Unit =
-        machine.ctrl = importCid(cid)
-      override def handleInactiveKey(machine: Machine, gkey: GlobalKey): Unit =
-        machine.ctrl = SEDamlException(DamlEContractKeyNotFound(gkey))
-      override def handleActiveKey(machine: Machine, cid: V.ContractId): Unit =
-        machine.returnValue = SContractId(cid)
+      override def handleInputKeyFound(cid: V.ContractId): SExpr =
+        importCid(cid)
+      override def handleKeyNotFound(gkey: GlobalKey): SEDamlException =
+        SEDamlException(DamlEContractKeyNotFound(gkey))
+      override def handleActiveKey(cid: V.ContractId): SEValue =
+        SEValue(SContractId(cid))
```

Comment:
We will paid a slight cost for using `SExpr` (and `machine.ctrl`)  instead of `SVaue` (and `machine.returnValue`). 
I do not think it is a big issue, as  underhanded the callback should be already very  costly.